docs/product/specs/permissions-revamp.md

```diff
@@ -0,0 +1,146 @@
+  * Several users have expressed interest towards using PostgreSQL roles, a few even mentioned that it's a blocker for them to start using Mathesar.
+  * Building our permission layer on top of PostgreSQL's permission system is future-proof and robust.
+- B. We should try to maintain feature parity with the flows we currently have.
+- C. Provide a rudimentary UX for ownerships and get users aquainted with the concept.
+
+## Features we intend to remove
+* We will remove the ability to share credentials across databases.
+  - We currently have this while creating/configuring a database.
+  - We will revisit this feature later down the road when we have requirements from users.
+
+## Terminologies
+* User refers to Mathesar user.
+* Admin refers to Mathesar admins.
+* Role refers to PostgreSQL role/user.
+
+## What we want the user to be able to do
+### Adding/configuring a database
+* A database should be unique. The same database cannot be added multiple times.
+  - This is a change from the existing setup where a database can be added multiple times with different connection credentials.
+* We'll have separate onboarding instructions/flows in the UI for pre-existing and new databases.
+
+#### Pre-existing database
+* Mathesar admins should be able to configure a pre-existing database, using the following details:
+  - DB server Host
+  - DB server Port
+  - Database name
+  - PostgreSQL role -> username and password
+* Mathesar admins should be able to optionally generate pre-defined roles for Editor, and Viewer, to partially satisfy Goal (B).
+  - This requires the PostgreSQL role used in the previous step to contain CREATEROLE attribute on the database server.
+  - This step is optional and Mathesar should function normally without these roles.
+  - We should make it clear that these roles & their privileges would only apply for new objects created via the Mathesar UI.
+  - For existing objects, we'll provide an option to iteratively provide privileges to these roles.
+    - This will however depend on ownerships of each object and can fail.
+  - We will no longer have the manager role as it would be the same as having admin access.
+    - Future consideration: We could provide a concept of database admins as a Mathesar level access control option per database. This will help maintain feature parity with the 'DB level admins' feature we currently have.
+#### New database
+* Mathesar admins should be able to create a new database on the internal database server by only entering the database name.
+  - This will use the same role utilized by the Django database.
+    - The user should be able to create new roles on the internal db and use them if they wanted to.
+  - The role used should contain both CREATEDATABASE and CREATEROLE privileges.
+* Mathesar admins should be able to create a new database on external servers, by specifying the same details as the 'pre-existing' database configuration.
```

Comment:
I know we've already discussed this, but this walkthrough makes the namespacing of a role under a database more obviously overcomplicated. Why can't the admin who has a connection to the external server already established (with sufficient privileges) just create a new DB by entering a database name?

It's also confusing for the user. They'll have to enter a previous database, as well as credentials that have privileges to connect to _that_ database (along with appropriate `CREATEDATABASE` privileges. If I were a user, I'd be trying to figure out:
- Why I had to enter an arbitrary preexisting database (You could enter any that the role in question has `CONNECT` privileges on)
- What, exactly, am I creating here?
  - the database (this one is obvious)
  - A new role?
  - A copy of the role?
Even more questions arise for a user who _does_ understand how roles, etc. work in PostgreSQL, since they'll be naturally confused as to why we're simulating a bunch of roles when there's actually only one.

After reading through this in its current form, I'm even more convinced that trying to namespace roles under databases in the users' minds is a complicating rather than simplifying abstraction.

Reply:
I updated an older version of the spec I had and forgot to remove this point. Thanks for pointing this out, I'll remove this.

Mathesar admins **should not** be able to create a new database on external servers.

We decided to remove the feature of being able to add a new database on external servers and sharing credentials. When we try to introduce it in the future, we can figure out a way of introducing the concept of database servers.

Reply:
Here's the commit https://github.com/mathesar-foundation/mathesar-wiki/pull/110/commits/189b77285da6da17d2179f28c83d6233d2d557af that should make this clear.